src/client-side-encryption/client_encryption.ts

```diff
@@ -541,13 +561,23 @@ export class ClientEncryption {
       }
     } = options;
 
+    const timeoutContext = this._timeoutMS
+      ? new CSOTTimeoutContext({
+          timeoutMS: this._timeoutMS,
+          serverSelectionTimeoutMS: this._client.options.serverSelectionTimeoutMS
+        })
+      : undefined;
+
     if (Array.isArray(encryptedFields.fields)) {
       const createDataKeyPromises = encryptedFields.fields.map(async field =>
         field == null || typeof field !== 'object' || field.keyId != null
           ? field
           : {
               ...field,
-              keyId: await this.createDataKey(provider, { masterKey })
+              keyId: await this.createDataKey(provider, {
+                masterKey,
+                timeoutContext: timeoutContext?.clone()
```

Comment:
Can we add a comment here explaining that we're cloning the timeoutContext to avoid concurrent use of timeout contexts which would lead to sharing the same timeout for server selection and connection checkout across different operations.

Reply:
Added it in!